tf_encrypted/layers/dense.py

```diff
@@ -37,25 +40,39 @@ def initialize(
         if initial_weights is None:
             initial_size = (self.in_features, self.out_features)
             initial_weights = np.random.normal(scale=0.1, size=initial_size)
-        if initial_bias is None:
-            initial_bias = np.zeros((1, self.out_features))
+        if initial_bias is not None:
+            self.bias = self.prot.define_private_variable(initial_bias)
 
         self.weights = self.prot.define_private_variable(initial_weights)
-        self.bias = self.prot.define_private_variable(initial_bias)
 
     def forward(self, x):
         self.layer_input = x
-        y = x.matmul(self.weights) + self.bias
+
+        if self.transpose_input:
+            self.layer_input = self.layer_input.transpose()
+        if self.transpose_weight:
+            self.weights = self.weights.transpose()
+
+        if self.bias:
+            y = x.matmul(self.weights) + self.bias
+        else:
+            y = x.matmul(self.weights)
         return y
 
     def backward(self, d_y, learning_rate):
         x = self.layer_input
+        if self.transpose_input:
+            self.layer_input = self.layer_input.transpose()
+        if self.transpose_weight:
+            self.weights = self.weights.transpose()
+
         d_x = d_y.matmul(self.weights.transpose())
 
         d_weights = x.transpose().matmul(d_y)
-        d_bias = d_y.reduce_sum(axis=0)
-
         self.weights.assign((d_weights * learning_rate).neg() + self.weights)
-        self.bias.assign((d_bias * learning_rate).neg() + self.bias)
+
+        if self.bias:
+            d_bias = d_y.reduce_sum(axis=0)
+            self.bias.assign((d_bias * learning_rate).neg() + self.bias)
```

Comment:
doesn't `assign` return an operation that has to be called for it to have any effect?

Reply:
hmm not sure, it was like that when Morgan was training the logistic regression and it seemed to work.

Reply:
yes it does: https://github.com/mortendahl/tf-encrypted/blob/3a55ca33247dad91199669a72c3e51511f7b698c/tf_encrypted/protocol/pond.py#L614

Reply:
why wouldn't we do `self.bias -= (d_bias * learning_rate)` here?

Reply:
Thanks for your comment @jvmancuso let me add this